src/CommandLine.cc

```diff
@@ -0,0 +1,445 @@
+            "                 default is: " APP_SHORTNAME ".\n"
+#if USE_WIN32_SERVICE
+            "       -r        Removes a Windows Service (see -n option).\n"
+#endif
+            "       -s | -l facility\n"
+            "                 Enable logging to syslog.\n"
+            "       -u port   Specify ICP port number (default: %d), disable with 0.\n"
+            "       -z        Create missing swap directories and then exit.\n"
+            "       -C        Do not catch fatal signals.\n"
+            "       -D        OBSOLETE. Scheduled for removal.\n"
+            "       -F        Don't serve any requests until store is rebuilt.\n"
+            "       -N        Master process runs in foreground and is a worker. No kids.\n"
+            "       --foreground\n"
+            "                 Master process runs in foreground and creates worker kids.\n"
+            "       --kid\n"
+            "                 A kid name passed to a worker Squid process.\n"
+            "                 Warning: do not use it directly since it is an internal option.\n"
+#if USE_WIN32_SERVICE
+            "       -O options\n"
+            "                 Set Windows Service Command line options in Registry.\n"
+#endif
+            "       -R        Do not set REUSEADDR on port.\n"
+            "       -S        Double-check swap during rebuild.\n"
+            "       -X        Force full debugging.\n"
+            "       -Y        Only return UDP_HIT or UDP_MISS_NOFETCH during fast reload.\n",
+            APP_SHORTNAME, CACHE_HTTP_PORT, DefaultConfigFile, CACHE_ICP_PORT);
+    exit(EXIT_FAILURE);
+}
+
+CommandLine::CommandLine(const int anArgc, char *anArgv[])
+    : execFile_(anArgv[0])
+{
+    parse(anArgc, anArgv);
+}
+
+void
+CommandLine::parse(int anArgc, char *anArgv[])
+{
+    int optIndex = 0;
+
+    // short options
+    const char *shortOpStr =
+#if USE_WIN32_SERVICE
+        "O:Vir"
+#endif
+        "CDFNRSYXa:d:f:hk:m::n:sl:u:vz?";
+
+    // long options
+    static struct option squidOptions[] = {
+        {"foreground", no_argument, 0,  ForegroundCode},
+        {"kid",        required_argument, 0, KidCode},
+        {"help",       no_argument, 0, 'h'},
+        {"version",    no_argument, 0, 'v'},
+        {0, 0, 0, 0}
+    };
+
+    int c;
+
+    while ((c = getopt_long(anArgc, anArgv, shortOpStr, squidOptions, &optIndex)) != -1) {
+        if (c == '?' || c == 'h') {
+            usage();
+            // unreacheable
+            break;
+        }
+        options.push_back(std::make_pair(c, optarg ? SBuf(optarg) : SBuf()));
+    }
+
+    // reserve space for --kid option (with argument) and nil termination pointer
+    argv_.clear();
+    argv_.reserve(anArgc + 3);
+    for (int i = 0; i < anArgc; ++i)
+        argv_.push_back(anArgv[i]);
+}
+
+const char **
+CommandLine::argv(const char *argv0, const char *kid)
+{
+    argv_[0] = argv0;
+
+    auto kidPos = std::find_if(argv_.begin(), argv_.end(), [](const char *opt)
```

Comment:
This is (a little incorrect) argv parsing. We already have code for doing that (correctly) elsewhere. If possible, we should not duplicate that functionality.

Reply:
We need to create a new argv for a kid. For the moment I had implemented two approaches:
1. Create it from scratch, filling with parsed arguments. We rejected due to implementation complexities.
2. Use the existing argv as a template, substituting/adding 'kids' argument and argv[0]. You reject it now either, saying that it is incorrect.
What do you suggest instead?

Reply:
I suggest using the CommandLine class as a C++ class, not a C struct with functions. CommandLine _is_ argv. If you need a copy of another argv, copy it!

* For kid argv creation, please see another [comment](https://github.com/squid-cache/squid/pull/176#discussion_r174279345) for a specific sketch.
* For the `--kid` value extraction, I would add a CommandLine::getOption(MyOptions_, MyOptions::optKid) or a similar method. That new method would iterate all options until it finds the requested one. The implementation would be similar to getNextOption() implementation but with a return statement instead of a callback call. The two methods may use some common private method or even a helper class to avoid code duplication. But I am starting to dictate code; there may be a better way to implement these two similar option iteration needs...

Reply:
> For kid argv creation, please see another comment with a specific sketch.

Your sketch looks good, but unfortunately, it does not scetch how that kidCommand.argv() should be implemented: that is what this question is about! My two above items were about such implementation, but both of them were rejected.

Reply:
I probably do not understand the problem you are trying to avoid or solve because IMO CommandLine::argv() is trivial. Here is its sketch:

```
X argv() const { return argv_.data(); }
...
std::vector<char*> argv_; ///< raw main() parameters, including argv[0]
```

... where X is whatever type exec*() and getopt_long() function(s) we use expect -- I did not risk typing it directly above. We might need a couple of such methods to return the right type (by casting data() -- it is all the same structure/bytes underneath).

We could also assert a positive argv_.size() before calling data(), but it is probably better to do so in the constructor instead -- `argv[0]` must always be present because the argv API requires a nil terminating member (at least).